[17.0][MIG] sale_coupon_criteria_order_based

mmequignon opened the pull request.

Commit:
[14.0][ADD] sale_coupon_criteria_order_based

francesco-ooops committed:
Added translation using Weblate (Italian)

francesco-ooops committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (10 of 10 strings)

Translation: sale-promotion-14.0/sale-promotion-14.0-sale_coupon_criteria_order_based
Translate-URL: https://translation.odoo-community.org/projects/sale-promotion-14-0/sale-promotion-14-0-sale_coupon_criteria_order_based/it/

pilarvargas-tecnativa committed:
[MIG] sale_coupon_criteria_order_based: Migration to version 15.0

TT38306

Commit:
[UPD] Update sale_coupon_criteria_order_based.pot

OCA-git-bot committed:
[UPD] README.rst

weblate committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: sale-promotion-15.0/sale-promotion-15.0-sale_coupon_criteria_order_based
Translate-URL: https://translation.odoo-community.org/projects/sale-promotion-15-0/sale-promotion-15-0-sale_coupon_criteria_order_based/

francesco-ooops committed:
Translated using Weblate (Italian)

Currently translated at 85.7% (6 of 7 strings)

Translation: sale-promotion-15.0/sale-promotion-15.0-sale_coupon_criteria_order_based
Translate-URL: https://translation.odoo-community.org/projects/sale-promotion-15-0/sale-promotion-15-0-sale_coupon_criteria_order_based/it/

OCA-git-bot committed:
[UPD] README.rst

Ivorra78 committed:
Added translation using Weblate (Spanish)

Ivorra78 committed:
Translated using Weblate (Spanish)

Currently translated at 100.0% (7 of 7 strings)

Translation: sale-promotion-15.0/sale-promotion-15.0-sale_coupon_criteria_order_based
Translate-URL: https://translation.odoo-community.org/projects/sale-promotion-15-0/sale-promotion-15-0-sale_coupon_criteria_order_based/es/

pilarvargas-tecnativa committed:
[FIX] sale_coupon_criteria_order_based: Add tagged "post_install" to tests

This test requires a CoA to be installed, it should be tagged "post_install"

TT46134

OCA-git-bot committed:
[BOT] post-merge updates

mmequignon committed:
sale_coupon_criteria_order_based: Pre-commit

mmequignon committed:
sale_coupon_criteria_order_based: Migration to 17.0

mmequignon committed:
sale_loyalty_criteria_order_based: rename sale_coupon_criteria_order_based

mmequignon committed:
sale_loyalty_criteria_order_based: Adapt to 17.0 odoo changes

mmequignon commented:
ping @francesco-ooops @Ivorra78 @pilarvargas-tecnativa 
could you please drop a review ?
Thanks!

francesco-ooops reviewed (changes requested):
- Please check whether documentation is still accurate (eg: there's no "**Sales/Products/Promotion Programs** menu" in v17)

- Domain button is not visible: 

![image](https://github.com/user-attachments/assets/a76ce63e-b013-4d21-ab42-2bf2fef56217)

- Also, is it correct that "Based on order" domain is both visible in promotion form and "conditional rules" popup?

mmequignon commented:
> * Please check whether documentation is still accurate (eg: there's no "**Sales/Products/Promotion Programs** menu" in v17)

You're right, I totally forgot to update the doc


>     * Domain button is not visible:

Yes, adding it right now


> ![image](https://private-user-images.githubusercontent.com/55489873/380192230-a76ce63e-b013-4d21-ab42-2bf2fef56217.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjk4NjUwMDksIm5iZiI6MTcyOTg2NDcwOSwicGF0aCI6Ii81NTQ4OTg3My8zODAxOTIyMzAtYTc2Y2U2M2UtYjAxMy00ZDIxLWFiNDItMmJmMmZlZjU2MjE3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDI1VDEzNTgyOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTExYzY0MDFhMDYzYWE4M2I2OWJhM2E3NDQxMmU2MjUzYWVjZDgzNjRkMDYwMmZlNzkzYzM0OWU2MmM2YjE2NTcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.XDQF2Q7WSAV-woFim93yFmokmgVVfZgOl1rbSziLD0o)
> 
>     * Also, is it correct that "Based on order" domain is both visible in promotion form and "conditional rules" popup?

Yes, and they are evaluated sequencially, which allows to have more complex rules.
1 domain which is common to all rules set on the program.
1 domain on the rule itself, which is specific to it.
With that we can imagine for a given program, the order should always be draft, 1 rule requires 10$ spent, the other a specific product to trigger the access to the reward.

francesco-ooops commented:
@mmequignon that's good, what about adding to the doc also this explanation about the presence of the field in both the promo and the condition rule to make more explicit how it's supposed to work?

mmequignon commented:
@francesco-ooops I updated the PR, can you please make another pass ?
Thanks !

francesco-ooops commented:
@mmequignon ping me after tests are fixed :)

mmequignon commented:
> @mmequignon ping me after tests are fixed :)

@francesco-ooops done

francesco-ooops commented:
@mmequignon mmm I don't see the field

![image](https://github.com/user-attachments/assets/4ad26ece-f0fc-4f5e-ab45-e1084e3986ed)

mmequignon commented:
> @mmequignon mmm I don't see the field
> 
> ![image](https://private-user-images.githubusercontent.com/55489873/380756156-4ad26ece-f0fc-4f5e-ab45-e1084e3986ed.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzAxMjYyNDUsIm5iZiI6MTczMDEyNTk0NSwicGF0aCI6Ii81NTQ4OTg3My8zODA3NTYxNTYtNGFkMjZlY2UtZjBmYy00ZjVlLWFiNDUtZTEwODRlMzk4NmVkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDI4VDE0MzIyNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTZiYzAyYWFhYThjYWI1ZjllMjA3NjA3YTZjNmM4YmE5ZjNkYWM0MDA1OTAwZjExMDExYTljYWQ5ZmFhMTVmNjYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Kko1h97kSXUg9wlhpldj586dLH-C9NjRcCxJ6miGE_Y)

That is because I added `base.group_no_one` (as it is done on product_domain) on the field.
you have to be have debug with assets enabled to see it

francesco-ooops commented:
Well, 

1) why?

2) that's the kind of stuff that makes functionals go to devs and ask "why don't I see this field?" and after their reply, "why isn't that in the documentation?"

I don't want to be rude, but our time is valuable too :)

mmequignon commented:
> Well,
> 
>     1. why?
> 
>     2. that's the kind of stuff that makes functionals go to devs and ask "why don't I see this field?" and after their reply, "why isn't that in the documentation?"
> 
> 
> I don't want to be rude, but our time is valuable too :)

I see your point, I just thought it would be better to stick to what was done for `product_domain`.
I'll drop this groups attribute.

pilarvargas-tecnativa commented on sale_loyalty_criteria_order_based/models/loyalty_program.py:
```diff
@@ -0,0 +1,34 @@
+class LoyaltyProgram(models.Model):
+    _inherit = "loyalty.program"
+
+    rule_order_domain = fields.Char(string="Based on Order", default="[]")
```
Where are you going to apply the criteria? At the rule level or at the program level?

grindtildeath replied:
Wouldn't it make more sense to use a M2o on `ir.filter` in order to make it reusable between different programs?

pilarvargas-tecnativa commented on sale_loyalty_criteria_order_based/models/loyalty_rule.py:
```diff
@@ -0,0 +1,35 @@
+class LoyaltyRule(models.Model):
+    _inherit = "loyalty.rule"
+
+    rule_order_domain = fields.Char(string="Based on Order", default="[]")
```
Where are you going to apply the criteria? At the rule level or at the program level?

francesco-ooops reviewed (approved):
Functional ok!

francesco-ooops commented:
just a suggestion: UX-wise, it would be useful if domain condition would be displayed in the "condition" card, in order to make all the rules explicit (as it happens for other fields)

![image](https://github.com/user-attachments/assets/aa765af0-cd67-4aaa-8564-34550d2b93a6)

mmequignon commented:
> just a suggestion: UX-wise, it would be useful if domain condition would be displayed in the "condition" card, in order to make all the rules explicit (as it happens for other fields)
> 
> ![image](https://private-user-images.githubusercontent.com/55489873/380782971-aa765af0-cd67-4aaa-8564-34550d2b93a6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzAxMzEzMzMsIm5iZiI6MTczMDEzMTAzMywicGF0aCI6Ii81NTQ4OTg3My8zODA3ODI5NzEtYWE3NjVhZjAtY2Q2Ny00YWFhLTg1NjQtMzQ1NTBkMmI5M2E2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDI4VDE1NTcxM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWFkMTQ5ZDcyMTI5OWQ4NzRkOTBiYjNiODI2Nzg5ZGQ4ZjE4MzYwZjc2MTFmOTJhZmM2NGU5ZWFhODhlM2ZkNmYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.2wOYw0TAdZiFP6GqfYLzvYQ1xHE-__ZX7WcRVmn8TCM)

I'll add this to the ROADMAP

grindtildeath reviewed (approved):
LGTM :+1:

pilarvargas-tecnativa reviewed (changes requested):
The criteria cannot be applied at rule level and at program level. You have to decide at which level it should be applied and follow the odoo flow to apply a promotion and respect the changes that have been made since version 16.0 to adapt the promotion modules. You can check the 16.0 branch and see how other modules have been migrated for reference.

grindtildeath committed:
[FIX] sale_loyalty_criteria_order_based: Do not evaluate False domain

In case a promotion is created without defining any domain on the UI,
an error message 'Invalid domain on rule' would appear.

Avoiding  evaluation of the field if its value is Falsish removes the
error.

grindtildeath committed:
fixup! [FIX] sale_loyalty_criteria_order_based: Do not evaluate False domain

rousseldenis commented:
/ocabot migration sale_coupon_criteria_order_based

grindtildeath commented:
@OCA/crm-sales-marketing-maintainers can you guys please reopen this one? :pray:
